src/docs/swift-code-generation.md

```diff
@@ -0,0 +1,70 @@
+# Swift code generation
+
+Certain parts of the Swift ABI are unsable, such as metadata, async context, dynamic dispatch thunks, and cannot be directly projected into .NET without Swift wrappers. These wrappers provide the flexibility to encapsulate unstable ABI parts into a stable calling convention.
+
+The main tradeoff is between added flexibility for supporting more unsable ABI scenarios and the increased cost of shipping and trimming. To minimize complexity, we propose generating Swift wrappers only when absolutely necessary, keeping them as thin as possible.
+
+Below, we outline the specific scenarios where Swift wrappers are required.
+
+## Enums
+
+Priority: High (required for StoreKit2 and SwiftUI)
+
+Discriminated unions are not supported in C#, and their memory layout in Swift is opaque, fragile, and unstable. To interact with Swift enums, a wrapper is necessary for payoload access and initialization. 
+
+[Implementation details](binding-enums.md)
+
+## Async
+Priority: High (required for StoreKit2 and SwiftUI)
+
+Async context in Swift is unstable and require thin wrappers for synhronous invocation. The wrapper will convert async functions into synchronous calls by using a Task as an async worker and a DispatchSemaphore to block until the operation completes. This approach encapsulates the unstable Swift async context, exposing only the stable wrapper interface to .NET.
```

Comment:
I would expect that async Swift APIs are going to be exposed as async methods in C#.

"Sync-over-async" is a well-known antipattern that results in poor performance. (It is possible that it is not a real problem for the set of Swift APIs that we are looking at the moment.)

Reply:
We want to offer a solution for async in StoreKit2: https://developer.apple.com/documentation/storekit/product/3851116-products

> "Sync-over-async" is a well-known antipattern that results in poor performance.

We would be able to support callconv requirements for the Swift async. I am concerned about the context the layout is unknown/unstable. To resolve this, we could use a think Swift wrapper to represent the context, but project it as `IAsyncStateMachine` in C#.

I will explore options and update this proposal accordingly.

Reply:
We just need to be able to create a C# `Task` on top of the underlying swift. There are many ways to do that on the C# side - we just need to find what we can get from Swift (for example, being able to register a callback which is called when the async operation is finished would be enough to build a `Task` out of it on the C# side).